### PR TITLE
improve page load performance of large amount urls

### LINKED ISSUE #3494: On Page reload monitors do not load
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### Description

I'm using the docker image on a Nvidia TX2. Ubuntu 22.04.  ARM64 with eMMC for storage
It should be noted I have 400+ monitors with groups. Storing data for 32 days
On page reload I get in the console log
```
index.560f0f4a.js:492 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'name')
    at bw.<anonymous> (index.560f0f4a.js:492:12765)
    at Sn.emit (index.560f0f4a.js:487:15555)
    at bw.emitEvent (index.560f0f4a.js:487:42368)
    at bw.onevent (index.560f0f4a.js:487:42171)
    at bw.onpacket (index.560f0f4a.js:487:41839)
    at Sn.emit (index.560f0f4a.js:487:15555)
    at index.560f0f4a.js:487:47862
```
nothing useful in docker logs 

On `docker restart <container id>`  uptime kuma will load monitors 

### 👟 Reproduction steps

1.) Use docker image,
2.) load monitors 
3.) Reload page
4.) Find monitors do not load and console log message will appear every minute

### 👀 Expected behavior

Monitors should load

### 😓 Actual Behavior

They do not

### 🐻 Uptime-Kuma Version

1.22.1

### 💻 Operating System and Arch

Ubuntu 22.04

### 🌐 Browser

Chrome Version 115.0.5790.110 (Official Build) (64-bit)

### 🐋 Docker Version

Version:           20.10.21

### 🟩 NodeJS Version

_No response_

### 📝 Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by vishalsabhaya**
@CommanderStorm @louislam 
1) fix loop to async
2) query n+1 problem

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Resolves https://github.com/louislam/uptime-kuma/issues/5129

testing results:
monitor url count: 415
current: 
cpu: 1  memory 2 gb: 1 min 44 sec

after changes

database: aws rds marinadb db.t3.micro(memory: 1gb, cpu: 2vcpu)

with docker run on mac:
cpu: 0.5  memory 1 gb: 21 sec
cpu:   1  memory 2 gb: 11 sec
cpu: 1.5  memory 3 gb: 10 sec
cpu:   2  memory 4 gb: 10 sec

with ESC Farget(Linux/ARM64):
cpu: 0.5  memory 1 gb: 31 sec
cpu: 1  memory 2 gb: 6 sec
cpu: 2  memory 4 gb: 5 sec
Fixes #3494 

## Type of change

- Performance improvement

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)
1cpu-2gbram

### CODE REVIEW

**Comment by vishalsabhaya** [comment]
@CommanderStorm @louislam 

I debug & fix performance issues for insert/update/delete/pause/resume monitor.

current:
   page load time + 500ms to 1 sec
   it always fetches the latest data from the database & builds json. i felt that it was extra time.
   
  after changes:
   < 700ms
   
   I just updated the target record with the front end instead of all records.

**Comment by mabed-fr** [comment]
Can you give me the docker command to test your PR? I encounter the problem with 1000 monitor (4vcpu / 16go RAM)

File: server/uptime-kuma-server.js
Comment (CommanderStorm) [inline server/uptime-kuma-server.js:230]: Please make sure that there are comments that explain uncommon/nonobvious logic.
There are a few other instances if this ^^

Here is a longer form expaination[^1]

[^1]: https://www.youtube.com/watch?v=Bf7vDBBOBUA

File: server/model/monitor.js
Comment (CommanderStorm) [inline server/model/monitor.js:79]: This is now nothing async more in here, right?
=> Please remove the async from this function and make sure that where this is called is not preloaded..

File: server/server.js
Comment (CommanderStorm) [inline server/server.js:1654]: I don't think these two need to be send in serial.
Have you checked what happens if you don't?

File: src/mixins/socket.js
Comment (CommanderStorm) [inline src/mixins/socket.js:159]: this is currently a no-op, right?

File: src/mixins/socket.js
Comment (CommanderStorm) [inline src/mixins/socket.js:155]: Why not `this.monitorList = data.list;`?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Sorry for the late review, I am currently quite stressed work-wise..

While the ~16x(!) performance improvement in this PR is great, I don't quite like how you implemented some parts.
(most parts are fine)

`preloadData` is quite clunky and does not have the 
=> please move this into an object that we pass to the function instead of the current way of passing a lookup-object into the function.
The current way is likely error prown and without tests whatsoever way in this part extremely risky.

Only passing an object also has the advantage of allowing better type-checking via adding what is passed in to the doc-comment.

I have also left some comments below

File: server/uptime-kuma-server.js
Comment (vishalsabhaya) [reply server/uptime-kuma-server.js:230]: @CommanderStorm 
thank you i will remove unnecessary comment

File: server/model/monitor.js
Comment (vishalsabhaya) [reply server/model/monitor.js:79]: @CommanderStorm 
understood. i will remove & test again

File: server/server.js
Comment (vishalsabhaya) [reply server/server.js:1654]: @CommanderStorm 
understand but it makes a huge difference on loading time. I already test this scenario & make a decision.

if you call one by one for all monitor. it take time.

It's core part of changes while loading page. then next preload data.

can you explain why you think not required to send in serial?

File: src/mixins/socket.js
Comment (vishalsabhaya) [reply src/mixins/socket.js:159]: @CommanderStorm 

i don't understand no-op.

do you means not in used?

we are using this.

let me explain why op is required.

current scenario what will happened in case of add/update/delete
lets assume 400 URLs

1) page-load: fetching **400urls** & render
2) add-monitor: time for add into database + fetching **401urls** from database + building monitor.json + update list this.monitorList = data.list
3) update-monitor: time for update into database + fetching **401urls** from database + building monitor.json + update list this.monitorList = data.list
4) delete-monitor: time for delete into database + fetching **400urls** from database + building monitor.json + update list this.monitorList = data.list

after adding op parameter
1) page-load:  fetching 400urls from database + building monitor.json + update list this.monitorList = data.list
2) add-monitor: time for add into database + fetching **1 url** from database + building monitor.json + update list this.monitorList = data.list
3) update-monitor: time for update into database + fetching **1 url** from database + building monitor.json + update list this.monitorList = data.list
4) delete-monitor: time for delete into database + fetching **1 url** from database + building monitor.json + update list this.monitorList = data.list

it will make huge difference

page load time will be same but add/update/delete time will be < 1 sec instead of 11 sec

File: server/model/monitor.js
Comment (vishalsabhaya) [reply server/model/monitor.js:1541]: @CommanderStorm 
understand let me double check with it.

**Comment by vishalsabhaya** [comment]
@CommanderStorm 

Thank you for your effert to review my pr. i am really appreciate your hard work for community.
I can't understand below comment can you elaborate?

> While the ~16x(!) performance improvement in this PR is great, I don't quite like how you implemented some parts.
> (most parts are fine)
> 
> preloadData is quite clunky
> => please move this into a single object per toJSON that we pass to the function instead of the current way of passing a lookup-object into the function.
> The current way is quite likely error prown and without tests whatsoever way in this part extremely risky.
> 
> Only passing an object also has the advantage of allowing better type-checking via adding what is passed in to the doc-comment.
>

**Comment by vishalsabhaya** [comment]
@CommanderStorm  @louislam 

I found 1 more thing. it will improve more seconds.

marked realtime dot is maximum 43. why we are fetching 100 records for each monitor? is there any reason?
currently I am testing with 834 monitor
limit 100 takes: 13 sec
limit 43 takes: 7 sec

image

image

File: server/server.js
Comment (CommanderStorm) [reply server/server.js:1654]: True, my logic was off on this one.

=> Let's just remove the inner `Promise.all` and just give all promises to the outer one.

**Comment by CommanderStorm** [comment]
>  is there any reason?

I can imagine two:

- What happens if you zoom out?
- What happens on high-dpi devices?

image

If we change that part, lets please do this in a second PR ^^

File: src/mixins/socket.js
Comment (CommanderStorm) [reply src/mixins/socket.js:159]: > i don't understand no-op.

I mean https://wikipedia.org/wiki/NOP_(code)

Honestly, `DELETE` being mingled in next to `LIST`, `UPDATE` and `ADD` is quite messy and has caused by some confusion in the review.
As noted by you, they should be different. I don't think that differentiating via `op` is good though. Adding this, means that the spagetti-code pile gets larger

=> It would be better from a reasoning about code perspective to remove the code and make this into different endpoints.
This also allows furhter (operation specific) optimisations for creation, update, list and deletion.
I like [googles guidelines](https://cloud.google.com/apis/design/standard_methods) for this.

Comment (vishalsabhaya) [reply src/mixins/socket.js:159]: @CommanderStorm 

sorry for being late reply, i was a little busy.

ultimately your approach to revert these changes as per google guidelines?
am I correct?
if yes then might be can not reduce time while insert/update/delete. because anyhow we can not reduce unnecessary processing.

correct me if my understanding is wrong.

**Comment by vishalsabhaya** [comment]
> > is there any reason?
> 
> I can imagine two:
> 
> * What happens if you zoom out?
> * What happens on high-dpi devices?
> 
> image
> 
> If we change that part, lets please do this in a second PR ^^

yes, you are correct. i didn't think about it. we will do in next pr. but we should restrict max span. because for detail information already provided graph.

File: src/mixins/socket.js
Comment (CommanderStorm) [reply src/mixins/socket.js:159]: Not revert, but rework.
Splitting this into separate socketio-endpoints seems like a good call.

Currently everything is muddled into the same list endpoint.
Switching functionality off with `op` seems like somewhat of a bodge..

Are we taking past each other?

File: server/uptime-kuma-server.js
Comment (vishalsabhaya) [reply server/uptime-kuma-server.js]: @CommanderStorm 
removed unnecessary commets

File: server/server.js
Comment (vishalsabhaya) [reply server/server.js]: @CommanderStorm 
moved to outer one. thank you for suggestion

File: server/model/monitor.js
Comment (vishalsabhaya) [reply server/model/monitor.js:1571]: @CommanderStorm 
I simplified logic. can you review one more time.

File: server/model/monitor.js
Comment (vishalsabhaya) [reply server/model/monitor.js]: @CommanderStorm 
i have removed async

**Comment by vishalsabhaya** [comment]
@CommanderStorm 
i seen something strange. did you notice ever?

i did testing with 800 records.

when I remove below line. all monitoring is working fine but current case most of monitor is stopped. monitoring data is not updating so monitor is not starting.
await sleep(getRandomInt(300, 1000));
image


startMonitors does not have await that's why?
image

**Comment by CommanderStorm** [comment]
Sorry, above comment does not quite parse for me.

As far as I understand that line, this just means that the monitors are started one after another with random intervals in between.

The way this is executed, starting monitors should happen after the server started listening, but must not have completely/partially completed before starting the `versionCheck`.
That this is not done in a Promise.all makes a [thundering heard problem](https://en.m.wikipedia.org/wiki/Thundering_herd_problem) less likely. The speed of starting them is constant though.

As far as I understand async js (feel free to correct), that sounds partially reasonable.
Changing the monitor startup to be smarter can be done, but let's separate this Into its own PR. ^^

File: src/mixins/socket.js
Comment (vishalsabhaya) [reply src/mixins/socket.js:159]: @CommanderStorm 
i understand your point.
Splitting this into separate socketio-endpoints seems like a good call.

let me start work on that. if get time today I will update pr. its last changes other changes already done

Comment (vishalsabhaya) [reply src/mixins/socket.js]: @CommanderStorm 
i updated as per your instructions. can you confirm?

**Comment by vishalsabhaya** [comment]
@CommanderStorm 
now all changes are done. can you review & guide me further? we should merge after review?

**Comment by vishalsabhaya** [comment]
@CommanderStorm 
is there any new changes? 
sorry to ping you again. i know you are very busy

**Comment by CommanderStorm** [comment]
Sorry, got back from vacation on monday and am still picking up the pieces work wise and had not time for such a large PR.
I can answer issues or merge small PRs on the train but for this one, I need an hour to look if everything works as intended.

**Comment by vishalsabhaya** [comment]
@CommanderStorm 
ahhh, understood. take your time.

File: server/uptime-kuma-server.js
Comment (CommanderStorm) [inline server/uptime-kuma-server.js]:
```suggestion
```

File: server/uptime-kuma-server.js
Comment (CommanderStorm) [inline server/uptime-kuma-server.js]:
```suggestion
        const result = {};
        monitorList.forEach(monitor => result[monitor.id] = monitor.toJSON(preloadData));
```

**Review by CommanderStorm — APPROVED** [review]
I have left a few nits to apply, but mostly LGTM 🎉



> [!NOTE]
> This PR is part of the [`v2.0` merge window](https://github.com/louislam/uptime-kuma/milestone/24) => see #4500 for the bugs that need to be addressed before we can ship this release ^^

<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>

File: server/server.js
Comment (louislam) [inline server/server.js:1686]: Just checking this pr. Is this code correct? Because the function returns nothing.

**Pull request merged**
